### ISSUE TITLE
HasMany relation incorrectly named when using relation_name_strategy = foreign_key

### ISSUE THREAD

**Issue by coatesap** [reliese/laravel#176]
When setting `'relation_name_strategy' => 'foreign_key'` in config/models.php, hasMany relationships are incorrectly named. 

For instance, a table called `authors` and a table called `books` The `books` table has a foreign key `author_id`, which points to `authors.id`. When the `Author` model is generated, it creates a relationship like:

```php
public function books_author()
{
    return $this->hasMany(Book::class);
}
```

Whereas it should name it as:

```php
public function books()
{
    return $this->hasMany(Book::class);
}
```

---

For reference, here is a sample migration for testing:
```php
Schema::dropIfExists('authors');
Schema::create('authors', function (Blueprint $table) {
    $table->increments('id');
});
Schema::dropIfExists('books');
Schema::create('books', function (Blueprint $table) {
    $table->increments('id');
    $table->unsignedInteger('author_id');
    $table->foreign('author_id')->references('id')->on('authors');
});
```

**Comment by CristianLlanos** [comment]
It's all about method name collisions. Let's say we have a table `people` and a table `books` with an `author` and `editor` relations to the `people` table. The generated models would be generated like these:
```php
class Book extends Model
{
	protected $table = 'books';

	public function author()
	{
		return $this->belongsTo(Person::class, 'author_id');
	}

	public function editor()
	{
		return $this->belongsTo(Person::class, 'editor_id');
	}
}

class Person extends Model
{
	protected $table = 'people';

	public function books_author()
	{
		return $this->hasMany(Book::class, 'author_id');
	}

	public function books_editor()
	{
		return $this->hasMany(Book::class, 'editor_id');
	}
}
``` 
How should we name `books_author()` and `books_editor()` so that they are pretty and contextualized at the same time? It might be possible to use `books` for the first method, but what about the other one?

----
Migration
```php
use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

class CreateBooksTable extends Migration
{

    public function up()
    {
        Schema::create('people', function (Blueprint $table) {
            $table->increments('id');
        });
        Schema::create('books', function (Blueprint $table) {
            $table->increments('id');
            $table->unsignedInteger('author_id');
            $table->foreign('author_id')
                ->references('id')
                ->on('people')
                ->onDelete('cascade');
            $table->unsignedInteger('editor_id');
            $table->foreign('editor_id')
                ->references('id')
                ->on('people')
                ->onDelete('cascade');
        });
    }

    public function down()
    {
        Schema::dropIfExists('books');
        Schema::dropIfExists('people');
    }
}
```

**Comment by coatesap** [comment]
@CristianLlanos - you're totally correct, they do need naming differently to avoid collisions where there is more than one foreign key to the same table. 

Something like `$person->books_where_editor` and `$person->books_where_author` might be slightly more understandable, but I suspect that most attempts to tweak the naming convention will not keep everyone happy. Also `$author->books_where_author` would still be left feeling unnecessary. 

The only way I could see this being substantially improved would be if we detected foreign key names that matched the foreign table name, and simplify the relationship naming in those instances. For instance, if we had a FK `books.author_id` that pointed to an `authors` table, then we could then have the more desirable `$author->books`. Other foreign keys that don't directly match the naming of the foreign table would be generated as they currently are (or with the "where" addition mentioned above).

How feasible/desirable do you think that would be?

**Comment by CristianLlanos** [comment]
I like the idea very much 🙌. It'd be a breaking change, so we could introduce this new approach in the next major version, v1.0.0. I think it would be very cool if you could make a the pull request with the improvements. Thank you.

**Comment by coatesap** [comment]
Just a quick update to say that I've got a rough version of this working now. I'll add unit tests etc before making a PR, but it does massively help to avoid relationship name collisions, meaning that many more relationships get generated for models, which were previously being suppressed.

**Comment by CristianLlanos** [comment]
That is amazing. I believe it will very much depend on the database schema and how it was modeled :). I also believe that it will be a very welcomed feature. Thanks.

**Comment by coatesap** [comment]
Fixed in #192 and merged into v1 branch.
